src/main.c

```diff
@@ -4752,17 +4791,34 @@ static void __tracer_destroy_base_dev(struct snap_device *dev){
 
 	if(dev->sd_base_dev){
 		LOG_DEBUG("freeing base block device");
-		elastio_snap_blkdev_put(dev->sd_base_dev);
-		dev->sd_base_dev = NULL;
+#ifndef HAVE_BDEV_OPEN_BY_PATH
+	elastio_snap_blkdev_put(dev->sd_base_dev);
+	dev->sd_base_dev = NULL;
+#else
+	elastio_snap_bdev_release(dev->sd_bd_handle);
+	dev->sd_bd_handle = NULL;
+	dev->sd_base_dev = NULL;
+#endif
 	}
 }
 
 static int __tracer_setup_base_dev(struct snap_device *dev, const char *bdev_path){
 	int ret;
-
 	//open the base block device
 	LOG_DEBUG("finding block device");
+#ifndef HAVE_BDEV_OPEN_BY_PATH
 	dev->sd_base_dev = elastio_snap_blkdev_get_by_path(bdev_path, FMODE_READ, NULL);
+#else
+	dev->sd_bd_handle = elastio_snap_bdev_open_by_path(bdev_path, FMODE_READ, NULL);
+	if (!IS_ERR(dev->sd_bd_handle)) {
+		dev->sd_base_dev = dev->sd_bd_handle->bdev;
+	} else {
+		ret = PTR_ERR(dev->sd_bd_handle);
+		dev->sd_base_dev = NULL;
+		LOG_ERROR(ret, "error finding block device '%s'", bdev_path);
+		goto error;
+	}
+#endif
```

Comment:
There are a lot of ifdev around this function. May be do some this:
```
struct res_blkdev {
   *sd_base_dev;
#ifdef HAVE_BDEV_OPEN_BY_PATH
   *sd_bd_handle;
#endif
};

// Incapsulate all defs in this function
res_blkdev* elastio_snap_blkdev_get_by_path(const char *path, fmode_t mode, void *holder)
{
#if defined HAVE_BDEV_OPEN_BY_PATH
    res_blkdev res{NULL, NULL};
	res.sd_bd_handle = bdev_open_by_path(path, mode, holder, NULL);
	if (IS_ERR(dev->sd_bd_handle)) {
		return NULL;
	}
	res.sd_base_dev = res.sd_bd_handle->bdev;
	return res;
#elif HAVE_BLKDEV_GET_BY_PATH_4
	return res_blkdev{blkdev_get_by_path(path, mode, holder, NULL), NULL};
#else
	return res_blkdev{blkdev_get_by_path(path, mode, holder), NULL};
#endif
  
}

// Incapsulate all releases
void elastio_snap_bdev_release(struct res_blkdev* blkdev)
{
#ifdef HAVE_BDEV_OPEN_BY_PATH
	bdev_release(blkdev->sd_bd_handle);
#elif HAVE_BLKDEV_PUT_1
//#if LINUX_VERSION_CODE < KERNEL_VERSION(2,6,28)
	blkdev_put(blkdev->sd_base_dev);
#elif defined HAVE_BLKDEV_PUT_HOLDER
	blkdev_put(blkdev->sd_base_dev, NULL);
#else
	blkdev_put(blkdev->sd_base_dev, FMODE_READ);
#endif
}
....


// Usage:
dev->res_base_dev = elastio_snap_blkdev_get_by_path(bdev_path, FMODE_READ, NULL);

void some_function(dev->res_base_dev->sd_base_dev);
..
elastio_snap_bdev_release(dev->res_base_dev);
```

Reply:
Refactored in a similar way